statedb/treeupdate.go

```diff
@@ -140,16 +147,78 @@ func (u *TreeUpdate) subTree(cfg *treeConfig) (treeUpdate *TreeUpdate, err error
 	return treeUpdate, nil
 }
 
-// SubTreeSingle returns a TreeUpdate of a singleton SubTree whose root is stored
-// in the leaf with `cfg.Key()`, and is parametrized by `cfg`.
-func (u *TreeUpdate) SubTreeSingle(cfg *SubTreeSingleConfig) (*TreeUpdate, error) {
-	return u.subTree(cfg.treeConfig())
+// DeepSubTree allows opening a nested subTree by passing the list of tree
+// configurations.
+func (u *TreeUpdate) DeepSubTree(cfgs []*TreeConfig) (treeUpdate *TreeUpdate, err error) {
+	tree := u
+	for _, cfg := range cfgs {
+		if tree, err = tree.SubTree(cfg); err != nil {
+			return nil, err
+		}
+	}
+	return tree, nil
+}
+
+// DeepGet allows performing a Get on a nested subTree by passing the list
+// of tree configurations and the key to get at the last subTree.
+func (u *TreeUpdate) DeepGet(cfgs []*TreeConfig, key []byte) ([]byte, error) {
+	tree, err := u.DeepSubTree(cfgs)
+	if err != nil {
+		return nil, err
+	}
+	return tree.Get(key)
+}
+
+// DeepAdd allows performing an Add on a nested subTree by passing the list
+// of tree configurations and the key, value to add on the last subTree.
+func (u *TreeUpdate) DeepAdd(cfgs []*TreeConfig, key, value []byte) error {
+	tree, err := u.DeepSubTree(cfgs)
+	if err != nil {
+		return err
+	}
+	return tree.Add(key, value)
+}
+
+// DeepSet allows performing a Set on a nested subTree by passing the list
+// of tree configurations and the key, value to set on the last subTree.
+func (u *TreeUpdate) DeepSet(cfgs []*TreeConfig, key, value []byte) error {
+	tree, err := u.DeepSubTree(cfgs)
+	if err != nil {
+		return err
+	}
+	return tree.Set(key, value)
+}
+
+// AsTreeView returns a read-only view of this subTree that fulfills the
+// TreeViewer interface.
+func (u *TreeUpdate) AsTreeView() TreeViewer {
+	return &treeUpdateView{u}
+}
+
+// treeUpdateView is a wrapper over TreeUpdate that fulfills the TreeViewer
+// interface.
+type treeUpdateView struct {
+	*TreeUpdate
+}
```

Comment:
note that treeUpdateView also has all of TreeUpdate's methods; you might not want this.

if you do not want it, another plus is that you can avoid an allocation and extra pointer by doing `type treeUpdateView TreeUpdate`, then you convert between them like `(*treeUpdateView)(tu)`.